jukebox42/Octoprint-PrusaMMU
# Initial MK4 support. Handle command detection

# I CANNOT STRESS THIS ENOUGH, DO NOT TRY THIS BRANCH IT'S NOT READY FOR USERS.

TODO:
- [x] multi print command detection
- [ ] Single print dialog support
- [ ] Single print command detection
- [x] Better error code handling
- [x] Extra commands (cut, eject)

This Branch aims to introduce support for the MK3.5/3.9/4 with an MMU.

Background:
When Prusa Updated the MMU for the MK4 they no longer send over the serial commands for the MMU. This means the information we had available is slimmer. On the flip, it's far less noisy in your input logs.

Single print mode is more difficult with the MK3.5+. Because there's no "MMU Single" profile, the sliced files don't contain a `Tx` command. This is what we use to pause the print, show the selection, and then insert before the M109 (wait for extruder) for the MK3.

Here's what I plan to try:

For the MK3:
- Look for `Tx` (tool change promp). 
- Remove it
- Show the prompt
- Wait for `M109 S` (wait for heater)
- Inject a tool command 

For the MK3.5:
- Look for `M104 S` (heat nozzle start). 
- DO NOT remove it
- Show the prompt
- Wait for `M109 S` (wait for heater)
- Inject a tool command AND the mmu follow up commands I stole from a multi slice

For the MK3.9/4:
- Look for `M104 S` (heat nozzle start). 
- DO NOT remove it
- Show the prompt
- Wait for `M569 S0 E` (set extruder speed)
- Inject a tool command AND the mmu follow up commands I stole from a multi slice
